# HG changeset patch
# User Rik <rik@octave.org>
# Date 1378940740 25200
#      Wed Sep 11 16:05:40 2013 -0700
# Node ID 95bfa04ab514737bd4f64560107b637923072e70
# Parent  f58ad514372ccf7df0a6adf8e5f2e5279ef199ae
Fix bug in colorbar demo #24.
set() is broken from deep within listener hierarchy.  Workaround it.

* scripts/plot/colorbar.m(update_colorbar_clim):  Set axis limits in listener
and then jiggle the axis position to force a redraw.

diff --git a/scripts/plot/colorbar.m b/scripts/plot/colorbar.m
--- a/scripts/plot/colorbar.m
+++ b/scripts/plot/colorbar.m
@@ -294,23 +294,42 @@ function update_colorbar_clim (hax, d, h
   if (isaxes (hax)
       && (isempty (gcbf ()) || strcmp (get (gcbf (), "beingdeleted"), "off")))
     clen = rows (get (get (hax, "parent"), "colormap"));
     cext = get (hax, "clim");
     cdiff = (cext(2) - cext(1)) / clen / 2;
     cmin = cext(1) + cdiff;
     cmax = cext(2) - cdiff;
 
+    hiax = get (hi, "parent");
     if (vert)
       set (hi, "ydata", [cmin, cmax]);
-      set (get (hi, "parent"), "ylim", cext);
+      set (hiax, "ylim", cext);
     else
       set (hi, "xdata", [cmin, cmax]);
-      set (get (hi, "parent"), "xlim", cext);
+      set (hiax, "xlim", cext);
     endif
+
+    ## FIXME: Setting xlim or ylim from within a listener callback
+    ##        causes the axis to change size rather than change limits.
+    ##        Workaround it by jiggling the position property which forces
+    ##        a redraw of the axis object.
+    ##
+    ## Debug Example:
+    ## Uncomment the line below.
+    ##   keyboard;
+    ## Now run the the following code.
+    ##   clf; colorbar (); contour (peaks ())
+    ## Once the keyboard command has been hit in the debugger try
+    ##   set (hiax, "ylim", [0 0.5]) 
+    pos = get (hiax, "position");
+    pos(1) += eps;
+    set (hiax, "position", pos);
+    pos(1) -= eps;
+    set (hiax, "position", pos);
   endif
 endfunction
 
 function update_colorbar_cmap (hf, d, hi, vert, init_sz)
   persistent sz = init_sz;
 
   if (ishandle (hf) && strcmp (get (hf, "type"), "figure")
       && (isempty (gcbf ()) || strcmp (get (gcbf (), "beingdeleted"), "off")))
@@ -324,17 +343,16 @@ function update_colorbar_cmap (hf, d, hi
       sz = clen;
       ## Also update limits on axis or there will be white gaps
       update_colorbar_clim (get (hi, "parent"), d, hi, vert);
     endif
   endif
 endfunction
 
 function update_colorbar_axis (h, d, cax, orig_props)
-
   if (isaxes (cax)
       && (isempty (gcbf ()) || strcmp (get (gcbf (), "beingdeleted"),"off")))
     loc = get (cax, "location");
     obj = get (h);
     obj.__cbar_hax__ = h;
     obj.position = orig_props.position;
     obj.outerposition = orig_props.outerposition;
     [pos, cpos, vertical, mirror] =  ...
